### PR TITLE
JP-3630: Add slit-specific processing metadata

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3630](https://jira.stsci.edu/browse/JP-3630)

Add some new informational keys to multi-slit data to make clear which slit-specific processing has taken place.  Affected data is NIRSpec MOS and FS; keys are added for wavecorr, pathloss, and barshadow steps.

Requires an stdatamodels change to add the new keys to the slit and spec datamodels: https://github.com/spacetelescope/stdatamodels/pull/431,
now merged.

Also, I noticed in passing a small bug in pathloss and corrected it here: the error propagation is incorrect when the step is inverted.  Error values are not currently used in steps that require the inversion, so there should be no impact to current data products. I added unit tests to verify the change as well as some basic coverage tests.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [x] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Comment by melanieclarke** [comment]
Initial regtests with stdatamodels branch:
https://github.com/spacetelescope/RegressionTests/actions/runs/13726415702

**Comment by melanieclarke** [comment]
Updated regression tests with stdatamodels on main, following PR merge there:
https://github.com/spacetelescope/RegressionTests/actions/runs/14045025650

File: docs/jwst/wavecorr/description.rst [hunk 1]
```diff
@@ -9,6 +9,11 @@ The wavelength correction (``wavecorr``) step in the
 assignments for NIRSpec fixed-slit (FS) and MOS point sources that are
 known to be off center (in the dispersion direction) in their slit.
 
+Upon successful completion of the step, the status keyword "S_WAVCOR"
+in the primary header is set to "COMPLETE".  For each SCI extension, the "WAVECOR"
+keyword is set to True if the slit was wavelength corrected (it is a point
+source) or False if it was not corrected (it is a point source).
```
Comment (emolter) [inline docs/jwst/wavecorr/description.rst:15]: both of these say "(it is a point source)" I think there's supposed to be a NOT in one?

File: jwst/barshadow/tests/test_barshadow_step.py [hunk 2]
```diff
@@ -30,17 +31,19 @@ def nirspec_mos_model():
...
-    yield im_ex2d
-    im_ex2d.close()
-    im_wcs.close()
-    im.close()
-    hdul.close()
+    return im_ex2d
+
+
+@pytest.fixture(scope="module")
+def nirspec_mos_model():
```
Comment (emolter) [inline jwst/barshadow/tests/test_barshadow_step.py:38]: out of curiosity: what is the benefit of having separated `create_nirspec_mos_model` from `nirspec_mos_model` here?

File: jwst/pathloss/pathloss.py [hunk 3]
```diff
@@ -430,14 +430,6 @@ def do_correction(
...
             output_model.meta.cal_step.pathloss = "SKIPPED"
             corrections = None
-        elif inverse:
-            log.warning("Use of inversion with NIS_SOSS is not implemented. Skipping")
-            output_model.meta.cal_step.pathloss = "SKIPPED"
-            corrections = None
-        elif source_type is not None:
-            log.warning("Forcing of source type with NIS_SOSS is not implemented. Skipping")
-            output_model.meta.cal_step.pathloss = "SKIPPED"
-            corrections = None
```
Comment (emolter) [inline jwst/pathloss/pathloss.py:440]: What was the impetus for removing these?  This seems not to be related to adding status keywords or to the error propagation bug, but I could be wrong about the latter

**Review by emolter — COMMENTED** [review]
This looks great Melanie, thanks for taking it on!  I added two questions and one small comment.  I'm not an expert in these steps, so I'll leave a more detailed review to someone who knows these better

File: jwst/barshadow/tests/test_barshadow_step.py (hunk 2 above)
Comment (melanieclarke) [reply jwst/barshadow/tests/test_barshadow_step.py:38]: I originally separated it because I wanted to be able to import `create_nirspec_mos_model` in the pathloss tests and use it directly.  Importing functions works better than importing fixtures.  I ended up making a separate function/fixture in the pathloss tests because it needed slightly different input conditions, but kept the separation between function and module-level fixture in case it was useful for future development to have a function that could be imported and called as needed.

File: jwst/pathloss/pathloss.py (hunk 3 above)
Comment (melanieclarke) [reply jwst/pathloss/pathloss.py:440]: They're not directly related -- it's a separate bug, probably from a bad conflict resolution long ago.  These lines are exact repeats of the lines immediately above them, so the code is unreachable.  I noticed it while checking test coverage and removed them.

Comment (emolter) [reply jwst/pathloss/pathloss.py:440]: whoops, ok I missed that they were exact duplicates of the above, thanks

**Comment by stscirij** [comment]
I think the MOS tests would benefit from having more than 1 slit: since some of the functionality is to add metadata for all slits, just testing slit[0] doesn't really show that anything other than the first slit is modified, which one might expect if only 1 slit were modified instead of all of them.

File: jwst/extract_1d/extract.py [hunk 4]
```diff
@@ -742,6 +742,9 @@ def copy_keyword_info(slit, slitname, spec):
         "quadrant",
         "slit_xscale",
         "slit_yscale",
+        "wavelength_corrected",
+        "pathloss_correction_type",
```
Comment (hayescr) [inline jwst/extract_1d/extract.py:746]: Should the pathloss_correction_type also be copied to the IFU x1d?  Currently it looks like it's not since the IFU runs through its own extraction.

**Review by hayescr — COMMENTED** [review]
Just one minor comment, otherwise this looks good to me.

File: jwst/extract_1d/extract.py (hunk 4 above)
Comment (hayescr) [reply jwst/extract_1d/extract.py:746]: Ah good point, I had only looked at the IFU cal files, but didn't check the cubes (which I assume this keyword is not propagated to, so it couldn't be easily passed to the x1d).  And yes, it's likely not that critical for IFU, regardless, (especially since the logs are now stored).

Happy to move forward with the PR as is.

**Review by hayescr — APPROVED** [review]
